File: AMENAPP/ChurchDiscoveryModels.swift
import Foundation
import CoreLocation

struct ChurchSearchIntent: Codable, Equatable {
    var denominationPreferences: [String]
    var traditionPreferences: [String]
    var worshipStyle: [String]
    var teachingStyle: [String]
    var communityNeeds: [String]
    var lifeStage: [String]
    var servicePreferences: [String]
    var groupNeeds: [String]
    var accessibilityNeeds: [String]
    var languagePreferences: [String]
    var distancePreference: String
    var mustHave: [String]
    var niceToHave: [String]
    var avoid: [String]
    var confidence: Double
    var needsClarification: Bool
    var clarifyingQuestion: String?
}

struct ChurchDiscoveryRequest: Codable, Equatable {
    var rawQuery: String
    var approximateLocation: ChurchDiscoveryLocation?
    var filters: [String: String] = [:]
}

struct ChurchDiscoveryLocation: Codable, Equatable {
    var latitude: Double
    var longitude: Double
    var label: String?

    init(latitude: Double, longitude: Double, label: String? = nil) {
        self.latitude = latitude
        self.longitude = longitude
        self.label = label
    }

    init(coordinate: CLLocationCoordinate2D, label: String? = nil) {
        self.latitude = coordinate.latitude
        self.longitude = coordinate.longitude
        self.label = label
    }
}

struct ChurchDiscoveryResponse: Codable {
    var searchId: String?
    var intent: ChurchSearchIntent
    var results: [ChurchDiscoveryResult]
}

struct ChurchDiscoveryResult: Codable, Identifiable, Equatable {
    var id: String
    var churchId: String?
    var googlePlaceId: String?
    var name: String
    var denomination: String?
    var address: String
    var latitude: Double?
    var longitude: Double?
    var distanceMiles: Double?
    var website: String?
    var phone: String?
    var googleMapsUri: String?
    var verifiedByAmen: Bool
    var matchScore: ChurchMatchScore
    var explanation: ChurchMatchExplanation
    var summary: ChurchDiscoverySummary
    var nextBestAction: String

    var asChurch: Church? {
        guard let latitude, let longitude else { return nil }
        return Church(
            canonicalChurchId: churchId ?? googlePlaceId ?? id,
            name: name,
            denomination: denomination ?? "Christian Church",
            address: address,
            distance: distanceMiles.map { String(format: "%.1f mi", $0) } ?? "Distance unavailable",
            distanceValue: distanceMiles ?? 0,
            serviceTime: "Service times not verified yet",
            phone: phone ?? "No phone available",
            coordinate: CLLocationCoordinate2D(latitude: latitude, longitude: longitude),
            website: website
        )
    }
}

struct ChurchMatchScore: Codable, Equatable {
    var overall: Double
    var confidence: Double
    var categories: ChurchMatchScoreCategories
}

struct ChurchMatchScoreCategories: Codable, Equatable {
    var intentMatch: Double
    var communityFit: Double
    var distanceTravelFit: Double
    var ministryGroupFit: Double
    var verifiedAmenData: Double
    var freshnessCompleteness: Double
    var accessibilityLanguageFit: Double
}

struct ChurchMatchExplanation: Codable, Equatable {
    var whyThisMayFit: [String]
    var possibleMismatch: [String]
    var verifiedFacts: [String]
    var missingInfo: [String]
    var sources: [String]
}

struct ChurchDiscoverySummary: Codable, Equatable {
    var thisMayFitBecause: String
    var checkThisFirst: String
    var bestNextStep: String
}

struct ChurchDiscoveryFilter: Codable, Equatable {
    var distanceMiles: Double?
    var denomination: String?
    var worshipStyle: String?
    var teachingStyle: String?
    var groupNeed: String?
    var accessibilityNeed: String?
    var language: String?
    var serviceTime: String?
}

struct ChurchSavedPreference: Codable, Equatable {
    var preferredDenominations: [String]
    var communityNeeds: [String]
    var worshipStyles: [String]
    var maxDistanceMiles: Double
}

enum ChurchCommunityNeed: String, Codable, CaseIterable, Identifiable {
    case youngAdults = "Young adults"
    case families = "Families"
    case bibleTeaching = "Bible teaching"
    case smallGroups = "Small groups"
    case worshipNight = "Worship night"
    case recovery = "Recovery"
    case newBelievers = "New believers"
    case traditional = "Traditional"
    case contemporary = "Contemporary"
    case onlineOptions = "Online options"

    var id: String { rawValue }
}

File: Backend/functions/src/churchDiscovery/churchDiscoveryModels.ts
export type DistancePreference = "nearby" | "within_10_miles" | "within_25_miles" | "online_ok" | "unspecified";

export type ChurchNextBestAction =
    "view_details"
    | "get_directions"
    | "visit_website"
    | "save"
    | "ask_berean"
    | "find_group"
    | "contact_church";

export interface UserLocationContext {
    latitude: number;
    longitude: number;
    label?: string | null;
}

export interface ChurchSearchIntent {
    denominationPreferences: string[];
    traditionPreferences: string[];
    worshipStyle: string[];
    teachingStyle: string[];
    communityNeeds: string[];
    lifeStage: string[];
    servicePreferences: string[];
    groupNeeds: string[];
    accessibilityNeeds: string[];
    languagePreferences: string[];
    distancePreference: DistancePreference;
    mustHave: string[];
    niceToHave: string[];
    avoid: string[];
    confidence: number;
    needsClarification: boolean;
    clarifyingQuestion: string | null;
}

export interface ChurchSearchRequest {
    rawQuery: string;
    approximateLocation?: UserLocationContext | null;
    filters?: Record<string, unknown>;
}

export interface GooglePlaceChurchCandidate {
    source: "google";
    placeId: string;
    displayName: string;
    formattedAddress: string;
    latitude: number | null;
    longitude: number | null;
    rating: number | null;
    userRatingCount: number | null;
    websiteUri: string | null;
    nationalPhoneNumber: string | null;
    regularOpeningHours: string[];
    photos: Array<{ name: string; widthPx?: number; heightPx?: number }>;
    googleMapsUri: string | null;
    businessStatus: string | null;
}

export interface AmenChurchProfileCandidate {
    source: "amen";
    churchId: string;
    name: string;
    denomination: string | null;
    traditionTags: string[];
    worshipStyleTags: string[];
    teachingStyleTags: string[];
    communityTags: string[];
    smallGroups: string[];
    ministries: string[];
    address: string;
    latitude: number | null;
    longitude: number | null;
    website: string | null;
    phone: string | null;
    googlePlaceId: string | null;
    hours: string[];
    photos: string[];
    accessibility: string[];
    verifiedByAmen: boolean;
    lastVerifiedAt: string | null;
    safetyStatus: string | null;
    sourceAttribution: string[];
    completeness: number;
}

export interface ChurchMatchScore {
    overall: number;
    confidence: number;
    categories: {
        intentMatch: number;
        communityFit: number;
        distanceTravelFit: number;
        ministryGroupFit: number;
        verifiedAmenData: number;
        freshnessCompleteness: number;
        accessibilityLanguageFit: number;
    };
}

export interface ChurchMatchExplanation {
    whyThisMayFit: string[];
    possibleMismatch: string[];
    verifiedFacts: string[];
    missingInfo: string[];
    sources: string[];
}

export interface ChurchSearchResult {
    id: string;
    churchId: string | null;
    googlePlaceId: string | null;
    name: string;
    denomination: string | null;
    address: string;
    latitude: number | null;
    longitude: number | null;
    distanceMiles: number | null;
    website: string | null;
    phone: string | null;
    googleMapsUri: string | null;
    verifiedByAmen: boolean;
    matchScore: ChurchMatchScore;
    explanation: ChurchMatchExplanation;
    summary: {
        thisMayFitBecause: string;
        checkThisFirst: string;
        bestNextStep: string;
    };
    nextBestAction: ChurchNextBestAction;
}

export interface ChurchDiscoveryAuditEvent {
    uid: string | null;
    event: string;
    searchId?: string;
    churchId?: string | null;
    googlePlaceId?: string | null;
    queryCategory?: string;
    createdAt?: FirebaseFirestore.FieldValue;
}

export const EMPTY_INTENT: ChurchSearchIntent = {
    denominationPreferences: [],
    traditionPreferences: [],
    worshipStyle: [],
    teachingStyle: [],
    communityNeeds: [],
    lifeStage: [],
    servicePreferences: [],
    groupNeeds: [],
    accessibilityNeeds: [],
    languagePreferences: [],
    distancePreference: "unspecified",
    mustHave: [],
    niceToHave: [],
    avoid: [],
    confidence: 0.35,
    needsClarification: false,
    clarifyingQuestion: null,
};
